### PR TITLE
hold_swap => is_held

### PR DESCRIPTION
**Pull request by oliver-sanders**
Addresses the `isHeld` point of #3223

~Quick POC ~branch for replacing the `hold_swap` logic for a `is_held` flag to see how much work is involved.

* Remove the "held" task state.
* Make "is_held" orthogonal to task status.
* Check task state by calling `itask.status()` (similar solution to https://github.com/cylc/cylc-flow/pull/3147/files#diff-e43061de4a331d2d734a6dd1a5ced381R1463).

Logical Change:

* All task states can be held, so you can have a held succeeded task.
* Holding a task is equivalent to telling Cylc not to issue any more submissions (will need to document as such).

Broken tests:

* ~tests/restart/21-task-elapsed.t~
* ~tests/hold-release/20-reset-waiting-output.t~
* ~tests/hold-release/13-ready-restart.t~

TODO:

* [x] Add tests
* [x] Fix tests
* [x] DB upgrader

Questions:

* [x] Does this seem sensible
* [ ] What about the retrying state (can this become `is_retrying` or merged in with `is_held` - see https://github.com/cylc/cylc-flow/issues/3223#issuecomment-511836616)  **follow up**

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Includes an appropriate entry in the release change log `CHANGES.md`.
- [x] No documentation update required for this change.

### CODE REVIEW

**Comment by oliver-sanders** [comment]
Almost there, three genuine test failures:

* tests/restart/21-task-elapsed.t
* tests/hold-release/20-reset-waiting-output.t
* tests/hold-release/13-ready-restart.t

**Comment by oliver-sanders** [comment]
The remaining two test failures seem to be genuine and quite difficult to shake off:

* tests/hold-release/20-reset-waiting-output.t
* tests/hold-release/13-ready-restart.t

**Comment by oliver-sanders** [comment]
At the moment this change has broken the logic which keeps the prerequisite brokering system in-line with manual resets.

* hold workflow
* reset X to succeeded
* reset X back to waiting
* release workflow
* downstream task Y now runs (despite X being waiting)

When a task is reset it's [upstream prerequisites are reset](https://github.com/cylc/cylc-flow/blob/master/cylc/flow/task_state.py#L442), I'm having a hard time hunting down the logic which is supposed to do this for downstream prerequisites (postrequisites).

* Task status is being reset as intended
* Task outputs are being reset as intended

The status of the offending prerequisite object after X has been reset to waiting:

```
Prerequisite(satisfied={('X', '1', 'succeeded'): 'satisfied naturally'}, _all_satisfied=True, target_point_strings=['1'], start_point=1, pre_initial_messages=[], conditional_expression=None, point=1)
```

(this is unchanged from before the reset [when X was succeeded])

**Comment by hjoliver** [comment]
@oliver-sanders - 

> When a task is reset it's upstream prerequisites are reset, I'm having a hard time hunting down the logic which is supposed to do this for downstream prerequisites (postrequisites).

We discussed this on Riot chat (and I just added to that conversation).  The way I'd put it is, for `foo => bar`:
1. reset `foo` to waiting, and its own prerequistes get set unsatisfied
1. reset `foo` to succeeded, and its own outputs get set to completed
1. reset `foo` back to waiting, and its own outputs get set back to unsatisfied

The prerequisites of `bar` *do not get reset in any case*, but:
- in case 2. they get satisfied by the normal dependency matching mechanism
- and in case 3. they remain satisfied because there is no mechanism (except manual reset of `bar`) to reverse already-satisfied prerequisites.

And finally, in the test `hold-release/20` the downstream task gets reset back to waiting only because all the manual state manipulation is done with the suite held, and on release we set held tasks back to their pre-hold state.

**Comment by hjoliver** [comment]
Hopefully that helps (at least to understand what's going on ... not entirely sure of the implications for this PR).

**Comment by oliver-sanders** [comment]
> the downstream task gets reset back to waiting only because all the manual state manipulation is done with the suite held

> not entirely sure of the implications for this PR

If this change breaks the logic It's possible there may be some convolution of the task-held and suite-held statuses in the task pool.

**Comment by oliver-sanders** [comment]
> not entirely sure of the implications for this PR

Fixed in a54ab0a!

**Comment by matthewrmshin** [comment]
No Travis CI for the latest change for some reason?

**Comment by oliver-sanders** [comment]
> No Travis CI for the latest change for some reason?

I'll mark the PR as ready for review (not quite there yet) I think that'll get Travis going.

**Comment by oliver-sanders** [comment]
De-conflicted.

I think this should now work, one test may fail (if so I might move it to flakytests):

* tests/hold-release/14-hold-kill.t

**Comment by hjoliver** [comment]
3 Travis kicks required to get one of the non-flaky chunks to pass. :angry:

**Comment by kinow** [comment]
>3 Travis kicks required to get one of the non-flaky chunks to pass. angry 

I wonder if that's Travis being unstable, or if we need to move a few more tests for the flaky group?

**Comment by hjoliver** [comment]
> > 3 Travis kicks required to get one of the non-flaky chunks to pass. angry
> 
> I wonder if that's Travis being unstable, or if we need to move a few more tests for the flaky group?

I'm not sure either.  If we had enough people/time, someone to spend an entire week (or more) making the test battery + Travis CI more friendly would be nice :thinking:

**Comment by hjoliver** [comment]
The flaky chunk is looking good so far though.

File: cylc/flow/task_job_mgr.py [hunk 1]
```diff
@@ -141,12 +141,13 @@ def poll_task_jobs(self, suite, itasks, poll_succ=True, msg=None):
         _poll_task_job_callback() executes one specific job.
         """
         to_poll_tasks = []
-        pollable_statuses = set([
-            TASK_STATUS_SUBMITTED, TASK_STATUS_RUNNING, TASK_STATUS_FAILED])
+        pollable_statuses = {
+            TASK_STATUS_SUBMITTED, TASK_STATUS_RUNNING, TASK_STATUS_FAILED
+        }
```
Comment (matthewrmshin) [inline cylc/flow/task_job_mgr.py]: :+1: Finally shaking off Python 2.6.

File: cylc/flow/task_pool.py [hunk 2]
```diff
@@ -725,10 +752,15 @@ def reload_taskdefs(self):
...
-                        TASK_STATUS_SUBMIT_RETRYING, TASK_STATUS_RETRYING,
-                        TASK_STATUS_HELD]:
+                if (
+                        itask.state(
+                            TASK_STATUS_WAITING,
+                            TASK_STATUS_QUEUED,
+                            TASK_STATUS_SUBMIT_RETRYING,
+                            TASK_STATUS_RETRYING,
+                        )
+                        or itask.state(is_held=True)
```
Comment (matthewrmshin) [inline cylc/flow/task_pool.py]: Can it just be `itask.state.is_held`?

Comment (oliver-sanders) [reply cylc/flow/task_pool.py]: Yeah, why not!

File: cylc/flow/task_pool.py [hunk 3]
```diff
@@ -1031,7 +1031,7 @@ def spawn_all_tasks(self):
                     itask.tdef.spawn_ahead
                     or itask.state(TASK_STATUS_EXPIRED)
                     or (
-                        itask.state(is_held=False)
+                        itask.state.is_held == False
```
Comment (matthewrmshin) [inline cylc/flow/task_pool.py]: Or just `not task.state.is_held`? (Assuming that `None` does not matter?)

File: cylc/flow/tests/test_rundb.py [hunk 4]
```diff
@@ -53,11 +57,117 @@ def test_select_task_job_sqlite_error(self):
...
-        self.mocked_connection.execute.side_effect = DatabaseError
+        self.mocked_connection.execute.side_effect = sqlite3.DatabaseError
 
         r = self.dao.select_task_job("it'll", "raise", "an error!")
         self.assertIsNone(r)
 
 
+def create_temp_db():
+    temp_db = mktemp()
+    conn = sqlite3.connect(temp_db)
```
Comment (matthewrmshin) [inline cylc/flow/tests/test_rundb.py]: For testing, can use `':memory:'` to create a database in RAM.

Comment (oliver-sanders) [reply cylc/flow/tests/test_rundb.py]: Need to save to disk so the DB can be re-opened by the Cylc DAO.

Comment (matthewrmshin) [reply cylc/flow/tests/test_rundb.py]: Cam that be mocked?

Comment (oliver-sanders) [reply cylc/flow/tests/test_rundb.py]: If you know how...

Comment (matthewrmshin) [reply cylc/flow/tests/test_rundb.py]: Otherwise, please make sure that the temporary file is cleared out after the test.

Comment (oliver-sanders) [reply cylc/flow/tests/test_rundb.py]: Not sure how to mock, easy to housekeep.

**Comment by hjoliver** [comment]
Kicking Travis for `tests/hold-release/14-hold-kill.t`. Test passes locally.  On Travis its another file not found issue: `cylc.flow.exceptions.SuiteServiceFileError: Couldn't get contact` which I think we see quite a lot of. I'm not sure if the specific tests are flaky or something else weird is happening on Travis CI.

**Comment by hjoliver** [comment]
> tests/hold-release/14-hold-kill.t
failed again :thinking:

**Comment by kinow** [comment]
```cylc.flow.exceptions.ClientError: Contact info not found for suite "cylctb-20190807T234059Z/hold-release/14-hold-kill", suite not running?```

This is the same kind of error that I started seeing in our functional tests. Not sure if it was due to any change in our code or Travis, but it's recent-ish I think.

**Comment by hjoliver** [comment]
And failed again, same test.

Direct link to T-CI after-fail.sh output: https://travis-ci.org/cylc/cylc-flow/jobs/568894700

First error is actually this:
```
==== /home/travis/cylc-run/cylctb-20190807T234059Z/hold-release/14-hold-kill/log/job/1/sleeper 02/job.err ====
Sending DEBUG MODE xtrace to job.xtrace
Fatal Python error: initsite: Failed to import the site module

Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site.py", line 703, in <module>
    main()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site.py", line 694, in main
    execsitecustomize()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site.py", line 548, in execsitecustomize
    import sitecustomize
  File "/home/travis/build/cylc/cylc-flow/.travis/sitecustomize.py", line 24, in <module>
    coverage.process_startup()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/control.py", line 1289, n process_startup
    cov.start()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/control.py", line 690, in start
    self._init()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/control.py", line 266, in _init
    concurrency=concurrency,
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/collector.py", line 107, in __init__
    self.origin = short_stack()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/debug.py", line 143, in short_stack
    stack = inspect.stack()[limit:skip:-1]
  File "/opt/python/3.7.1/lib/python3.7/inspect.py", line 1511, in stack
    return getouterframes(sys._getframe(1), context)
  File "/opt/python/3.7.1/lib/python3.7/inspect.py", line 1488, in getouterframes
    frameinfo = (frame,) + getframeinfo(frame, context)
  File "/opt/python/3.7.1/lib/python3.7/inspect.py", line 1458, in getframeinfo
    filename = getsourcefile(frame) or getfile(frame)
  File "/opt/python/3.7.1/lib/python3.7/inspect.py", line 696, in getsourcefile
    if getattr(getmodule(object, filename), '__loader__', None) is not None:
  File "/opt/python/3.7.1/lib/python3.7/inspect.py", line 725, in getmodule
    file = getabsfile(object, _filename)
  File "/opt/python/3.7.1/lib/python3.7/inspect.py", line 709, in getabsfile
    return os.path.normcase(os.path.abspath(_filename))
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/posixpath.py", line 376, in abspath
    cwd = os.getcwd()
FileNotFoundError: [Errno 2] No such file or directory
```

**Comment by kinow** [comment]
Hmm, maybe the coverage script that uses the site-customization trick in Python is trying to use an invalid directory (deleted, incorrect, etc)

**Comment by oliver-sanders** [comment]
> Kicking Travis for tests/hold-release/14-hold-kill.t. Test passes locally.

I have managed to get it to fail locally with a busy system (and only after several attempts), simple solution, dump it in flakytests?

File: cylc/flow/tests/test_rundb.py (hunk 4 above)
Comment (hjoliver) [reply cylc/flow/tests/test_rundb.py]: nice

**Comment by matthewrmshin** [comment]
(Just kicked Travis CI again, noting failures of `hold-release/12` and `cli/02`. I can see why the hold-release tests are so sensitive to time. I'll have a look at them soon.)

File: cylc/flow/tests/test_rundb.py (hunk 4 above)
Comment (kinow) [reply cylc/flow/tests/test_rundb.py]: I think a temporary file can be used. @dwsutherland 's PR has some new test utility code that handles something similar (though I haven't followed the whole discussion here, so forgive me if out of context): https://github.com/cylc/cylc-flow/pull/3202/files#diff-8222eb14a16268057dbc4ac79670fa3fR109-R113

(in David's PR, there is a base class for tests, so the suite DB is created on `setUp`, and destroyed on `tearDown`)

**Comment by oliver-sanders** [comment]
`flakytests/hold-release/14-hold-kill.t` is now failing reliably on Travis but passing reliably on my system even when put under load:

```bash
for i in $(seq 1 10); do
    sleep 1  # due to ctb test naming convention
    etc/bin/run-functional-tests.sh tests/hold-release/14-hold-kill.t &
done
```

Will see if bc84e23 helps.

**Comment by oliver-sanders** [comment]
> Will see if bc84e23 helps.

It did!

Tests passing!

**Pull request merged**